go/cmd/server/logging/logging.go

```diff
@@ -0,0 +1,384 @@
+	"github.com/apache/arrow/go/v8/arrow/memory"
+	"github.com/feast-dev/feast/go/internal/feast"
+	"github.com/feast-dev/feast/go/internal/feast/model"
+	"github.com/feast-dev/feast/go/protos/feast/serving"
+	"github.com/feast-dev/feast/go/protos/feast/types"
+	gotypes "github.com/feast-dev/feast/go/types"
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+const DEFAULT_LOG_FLUSH_INTERVAL = 100 * time.Millisecond
+const DEFAULT_LOG_INSERT_TIMEOUT = 20 * time.Millisecond
+
+type Log struct {
+	// Example: val{int64_val: 5017}, val{int64_val: 1003}
+	EntityValue []*types.Value
+
+	FeatureValues   []*types.Value
+	FeatureStatuses []serving.FieldStatus
+	EventTimestamps []*timestamppb.Timestamp
+	RequestContext  map[string]*types.RepeatedValue
+	RequestId       string
+}
+
+type MemoryBuffer struct {
+	featureService *model.FeatureService
+	logs           []*Log
+}
+
+type LoggingService struct {
+	memoryBuffer      *MemoryBuffer
+	logChannel        chan *Log
+	fs                *feast.FeatureStore
+	offlineLogStorage OfflineLogStorage
+	logInsertTTl      time.Duration
+	logFlushInterval  time.Duration
+}
+
+func NewLoggingService(fs *feast.FeatureStore, logChannelCapacity int, featureServiceName string, enableLogProcessing bool) (*LoggingService, error) {
+	var featureService *model.FeatureService = nil
+	var err error
+	if fs != nil {
+		featureService, err = fs.GetFeatureService(featureServiceName)
+		if err != nil {
+			return nil, err
+		}
+
+	}
+
+	loggingService := &LoggingService{
+		logChannel: make(chan *Log, logChannelCapacity),
+		memoryBuffer: &MemoryBuffer{
+			logs:           make([]*Log, 0),
+			featureService: featureService,
+		},
+		fs:               fs,
+		logInsertTTl:     DEFAULT_LOG_INSERT_TIMEOUT,
+		logFlushInterval: DEFAULT_LOG_FLUSH_INTERVAL,
+	}
+
+	if fs != nil {
+		offlineLogStorage, err := NewOfflineStore(fs.GetRepoConfig())
+		loggingService.offlineLogStorage = offlineLogStorage
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Start goroutine to process logs
+	if enableLogProcessing {
+		go loggingService.processLogs()
+
+	}
+	return loggingService, nil
+}
+
+func (s *LoggingService) EmitLog(l *Log) error {
+	select {
+	case s.logChannel <- l:
+		return nil
+	case <-time.After(s.logInsertTTl):
+		return fmt.Errorf("could not add to log channel with capacity %d. Operation timed out. Current log channel length is %d", cap(s.logChannel), len(s.logChannel))
+	}
+}
+
+func (s *LoggingService) processLogs() {
+	// start a periodic flush
+	// TODO(kevjumba): set param so users can configure flushing duration
+	ticker := time.NewTicker(s.logFlushInterval)
+	defer ticker.Stop()
+
+	for {
+		s.HandleLogFlushing(ticker)
+	}
+}
+
+// Select that eitheringests new logs that are added to the logging channel, one at a time to add
+// to the in memory buffer or flushes all of them synchronously to the OfflineStorage on a time interval.
+func (s *LoggingService) HandleLogFlushing(t *time.Ticker) {
```

Comment:
This function does not only flushing but also pulls from logchannel and fills the buffer

Reply:
Changed the name again.